pandas/core/groupby/groupby.py

```diff
@@ -4941,6 +4948,33 @@ def shift(
         else:
             axis = 0
 
+        if is_list_like(periods):
+            # periods is not necessarily a list, but otherwise mypy complains.
+            periods = cast(list, periods)
+            if axis == 1:
+                raise ValueError(
+                    "If `periods` contains multiple shifts, `axis` cannot be 1."
+                )
+            if len(periods) == 0:
+                raise ValueError("If `periods` is an iterable, it cannot be empty.")
+            from pandas.core.reshape.concat import concat
+
+            shifted_dataframes = []
+            for period in periods:
+                if not isinstance(period, int):
+                    raise TypeError(
+                        f"Periods must be integer, but {period} is {type(period)}."
+                    )
+                shifted_dataframes.append(
+                    DataFrame(
+                        self.shift(
```

Comment:
This structure means the overhead (e.g. argument checking) is called for each iteration of the for loop. I think a better structure would be:

```
if not is_list_like(periods):
    periods = [periods]

# possibly more setup

shifted_dataframes = []
for period in periods:
    # Majority of current method in here
    ...
result = concat(shifted_dataframes)
```

Reply:
I like that less because to me the multiple periods scenario is a special case. After all, what we are doing is, we're looping over the periods and calling `.shift` again for each period. (It's a mini recursion)

So as I stated above, to me the more logical way to handle this code would be to factor out the special case (periods is list like).

You are correct that this is inefficient because checks are done twice, _but_ only one check is independent of the value of `period` and it's a simple check (`if freq is not None and fill_value is not lib.no_default:`). The others are downstream of checks of, e.g., whether `periods` is 0.

Reply:
Actually I think what should be done is, that check should be moved before the check for what periods is so that we fail as soon as possible (shorter stack trace).

Reply:
It looks to me like the call to grouper.group_info, len(ids), attribute access with `_obj_with_exclusions` all only need to happen once.

Reply:
🤦 I was looking at pandas/core/frame.py, not this one.